### PR TITLE
fix high CPU usage #1481

### PR DESCRIPTION
**Pull request by metacodes**
See detail discussion in #1437 .

### CODE REVIEW

File: src/logic/Application.swift [hunk 1]
```diff
@@ -158,20 +158,23 @@ class Application: NSObject {
...
-                try self.axUiElement!.subscribeToNotification(axObserver, notification, {
-                    DispatchQueue.main.async { [weak self] in
-                        guard let self = self else { return }
-                        // some apps have `isFinishedLaunching == true` but are actually not finished, and will return .cannotComplete
-                        // we consider them ready when the first subscription succeeds, and list their windows again at that point
-                        if !self.isReallyFinishedLaunching {
-                            self.isReallyFinishedLaunching = true
-                            self.observeNewWindows()
+        // we only need to subscribe to those apps which will have windows
+        if AXUIElement.isManageable(runningApplication) {
```
Comment (lwouis) [inline src/logic/Application.swift:162]: This check is a bit late. Better not even create the `Application` instance, if we think it will not spawn any window. There is already such a method to filter apps here: https://github.com/lwouis/alt-tab-macos/blob/02590579df936d6a2124707e12d76865636fe23f/src/logic/Applications.swift#L104

You could add your new filtering logic there.

Comment (metacodes) [reply src/logic/Application.swift:162]: Thank you for the suggestion. I found a compensation method to solve this problem. The initial technical attempts have been made, and I'll implement the code for this later. I'm a bit busy these two weeks.

File: src/api-wrappers/AXUIElement.swift [hunk 2]
```diff
@@ -247,6 +247,80 @@ extension AXUIElement {
...
+        "com.apple.TextInputSwitcher"
+    ])
+
+    /**
+     *  Returns a Bool indicating whether or not the application will have windows.
+     *
+     *  @return true if the application will have windows and false otherwise.
+     */
+    static func isManageable(_ runningApp: NSRunningApplication) -> Bool {
+        guard let bundleIdentifier = runningApp.bundleIdentifier else {
```
Comment (lwouis) [inline src/api-wrappers/AXUIElement.swift:278]: we have cases like the android emulator where the app has no `bundleIdentifier` but is still app with windows (see https://github.com/lwouis/alt-tab-macos/blob/02590579df936d6a2124707e12d76865636fe23f/src/logic/Applications.swift#L127)

File: src/api-wrappers/AXUIElement.swift [hunk 3]
```diff
@@ -247,6 +247,80 @@ extension AXUIElement {
...
+        "com.apple.PowerChime",
+        "com.apple.WebKit.Networking",
+        "com.apple.WebKit.WebContent",
+        "com.apple.WebKit.GPU",
+        "com.apple.FollowUpUI",
+        "com.apple.controlcenter",
+        "com.apple.SoftwareUpdateNotificationManager",
+        "com.apple.TextInputMenuAgent",
+        "com.apple.TextInputSwitcher"
+    ])
```
Comment (lwouis) [inline src/api-wrappers/AXUIElement.swift:270]: I don't think it's a good idea to hardcode specific `bundleIdentifier`s. If these start creating windows in further macOS releases, then AltTab will have to be updated to show them. Also how did you build that list? Maybe it's missing processes from previous macOS versions you don't use, or maybe (actually it's certain) different processes will run on your mac depending on your System Preferences, if you have a network printer, external monitor, keyboard, if your phone does AirPlay with the mac, etc.

I think we should find more flags or properties of apps that give strong hints that they will not spawn windows. Also retrying less often than 10ms is already a large workaround the CPU issue

File: src/api-wrappers/AXUIElement.swift [hunk 4]
```diff
@@ -21,7 +21,7 @@ func retryAxCallUntilTimeout_(_ group: DispatchGroup?, _ timeoutInSeconds: Doubl
     } catch {
         let timePassedInSeconds = Double(DispatchTime.now().uptimeNanoseconds - startTime.uptimeNanoseconds) / 1_000_000_000
         if timePassedInSeconds < timeoutInSeconds {
-            BackgroundWork.axCallsQueue.asyncAfter(deadline: .now() + .milliseconds(10)) {
+            BackgroundWork.axCallsQueue.asyncAfter(deadline: .now() + .milliseconds(1000)) {
```
Comment (lwouis) [inline src/api-wrappers/AXUIElement.swift:24]: I think a value of `250ms` may be a better comprise: 10ms is very aggressive for a responsive UX but at the cost of CPU. 1s is easy on the CPU, but a bit laggy. I think `250ms` is similar to human delay in processing changes on screen. So maybe a better compromise. See https://humanbenchmark.com/tests/reactiontime

**Comment by metacodes** [comment]
@lwouis I use new method to aovid subcribing to applications withou windows. See details in the code comments.

File: src/logic/Applications.swift [hunk 5]
```diff
@@ -104,7 +105,30 @@ class Applications {
     private static func isActualApplication(_ app: NSRunningApplication) -> Bool {
         // an app can start with .activationPolicy == .prohibited, then transition to != .prohibited later
         // an app can be both activationPolicy == .accessory and XPC (e.g. com.apple.dock.etci)
-        return (isNotXpc(app) || isAndroidEmulator(app)) && !app.processIdentifier.isZombie()
+        return (isNotXpc(app) || isAndroidEmulator(app)) && !app.processIdentifier.isZombie() && isAnWindowApplication(app)
```
Comment (lwouis) [inline src/logic/Applications.swift:108]: If your comment below is correct (that apps that trigger a `didActivateApplicationNotification` event must have windows), then all these checks are no longer useful for these apps, no?

I imagine that they would only be useful for the `wasLaunchedBeforeAltTab == true` apps. So maybe we should add that boolean check first (left-side), to avoid the more expansive checks in the case where it's `wasLaunchedBeforeAltTab == false`, no?

File: src/logic/events/WorkspaceEvents.swift [hunk 6]
```diff
@@ -14,11 +14,30 @@ class WorkspaceEvents {
...
+    }
+
+    // We add apps when we receive a didActivateApplicationNotification notification, not when we receive an apps launched, because any app will have an apps launched notification.
+    // But we are only interested in apps that have windows. We think that since an app can be activated, it must have a window, and subscribing to its window event makes sense and is likely to work, even if it requires multiple retries to subscribe.
+    // I'm not very sure if there is an edge case, but so far testing down the line has not revealed it.
+    // When we receive the didActivateApplicationNotification notification, NSRunningApplication.isActive=true, even if the app is not the frontmost window anymore.
+    // If we go to add the application when we receive the message of apps launched, at this time NSRunningApplication.isActive may be false, and try axUiElement.windows() may also throw an exception.
+    // For those background applications, we don't receive notifications of didActivateApplicationNotification until they have their own window. For example, those menu bar applications.
```
Comment (lwouis) [inline src/logic/events/WorkspaceEvents.swift:33]: If this is true, then your approach here to listen to `.didActivateApplicationNotification` is genius! Did you think or it yourself or was it used by other projects?

Also, if I summarize this change, it seems you found an API that gives us properly filtered apps (only apps with windows), but that's only for apps launched *after* AltTab is started. For apps launched before AltTab is started, we still have to filter out processes with `activationPolicy`, `XPC`, etc. But luckily, apps launched before AltTab should not throw an exception on `.windows()`.

And then you mention the edge-case of an app launching at the same time as AltTab, which would be annoying. This is actually a very important edge-case, unfortunately. That's what happens when you log into macOS. All the startup apps launch at once, concurrently. This means AltTab is launched, and all sorts of app at the very same time.

I guess for that scenario, `didActivateApplicationNotification` will not help us, and we will need to rely on the other existing checks (XPC, etc), and with the new `250ms` delay, it may be an acceptable situation?

Am I understanding correctly the whole approach?

Comment (metacodes) [reply src/logic/events/WorkspaceEvents.swift:33]: > Did you think or it yourself or was it used by other projects?

At first I was thinking that when we open an application, as long as they show a window, then a change in window switching is bound to happen, and if we can find a way to observe this change, we will be able to tell that the program is equipped with a window. Then I used some window management software, such as automatic window layout APP, like AppGrid, Rectangle, and found that they can observe this change very well, so I went to study their code, and then found in Rectangle's code that it subscribes to `.didActivateApplicationNotification` notifications. I was curious, so I wrote some test code to see the effect. Turns out it worked and was able to solve half of the problem, the other half still had to be figured out.

> And then you mention the edge-case of an app launching at the same time as AltTab, which would be annoying. This is actually a very important edge-case, unfortunately. That's what happens when you log into macOS. All the startup apps launch at once, concurrently. This means AltTab is launched, and all sorts of app at the very same time.

That's my concern, and I still need to do some testing on this situation.

> I guess for that scenario, didActivateApplicationNotification will not help us, and we will need to rely on the other existing checks (XPC, etc), and with the new 250ms delay, it may be an acceptable situation?

This edge case only happens when `wasLaunchedBeforeAltTab=true`, at which point we go to the `axUiElement.windows()` test. If it happens that the program takes a long time to load, then `axUiElement.windows()` will most likely throw an exception, and at that point we do need to detect it by other means (e.g. XPC, etc), otherwise we will lose the window. Of course, reopening the program that lost the window, AltTab will capture its window without any problem.

Comment (metacodes) [reply src/logic/events/WorkspaceEvents.swift:33]: For that edge-case, we probably don't need to restart the applications that are missing windows, we can do a check when we get the `.didActivateApplicationNotification` notification, and if the application is not in the AltTab list of existing windows, then we add it.

**Comment by metacodes** [comment]
@lwouis I have submitted a new commit for fixing #1324 and some minor improvements to full-screen thumbnails.

File: src/logic/Applications.swift [hunk 7]
```diff
@@ -115,7 +124,13 @@ class Applications {
...
             // I'm not quite sure if this happens, but even if it does, then after restarting this application, AltTab captures its window without any problem. I think this happens rarely.
             let allWindows = CGWindow.windows(.optionAll)
-            guard let winApp = (allWindows.first { app.processIdentifier == $0.ownerPID() && $0.isNotMenubarOrOthers() && $0.id() != nil && $0.title() != nil}) else {
+            guard let winApp = (allWindows.first { app.processIdentifier == $0.ownerPID()
+                    && $0.isNotMenubarOrOthers()
+                    && $0.id() != nil
+                    && $0.bounds() != nil
+                    && CGRect(dictionaryRepresentation: $0.bounds()!)!.width > 0
+                    && CGRect(dictionaryRepresentation: $0.bounds()!)!.height > 0
+            }) else {
```
Comment (lwouis) [inline src/logic/Applications.swift:133]: As I said in the other ticket, you're re-implementing a mirror to https://github.com/lwouis/alt-tab-macos/blob/b668928c89dd256b39084d47b523ea882ebf6833/src/api-wrappers/AXUIElement.swift#L76 here. This will not be good enough. AltTab will thus show windows it shouldn't show.

File: src/api-wrappers/PrivateApis.swift [hunk 8]
```diff
@@ -122,10 +122,15 @@ func CGSCopyWindowsWithOptionsAndTags(_ cid: CGSConnectionID, _ owner: UInt32, _
...
 
 // adds the provided windows to the provided spaces
-// * macOS 10.10+
+// * macOS 10.10-12.2
 @_silgen_name("CGSAddWindowsToSpaces")
 func CGSAddWindowsToSpaces(_ cid: CGSConnectionID, _ windows: NSArray, _ spaces: NSArray) -> Void
 
+// Move the given windows (CGWindowIDs) to the given space (CGSSpaceID)
+// * macOS 10.10+
+@_silgen_name("CGSMoveWindowsToManagedSpace")
```
Comment (lwouis) [inline src/api-wrappers/PrivateApis.swift:131]: Does this work exactly like `CGSAddWindowsToSpaces`? I remember testing almost every API from https://github.com/phracker/MacOSX-SDKs/blob/10dd4868459aed5c4e6a0f8c9db51e20a5677a6b/MacOSX10.10.sdk/System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics.tbd which had the word "space" or "window" in it. I'm surprised that I missed this one. Are you sure it works in every scenario (e.g. multiple monitors, minimized window, hidden app, etc)?

File: src/api-wrappers/PrivateApis.swift [hunk 9]
```diff
@@ -122,10 +122,15 @@ func CGSCopyWindowsWithOptionsAndTags(_ cid: CGSConnectionID, _ owner: UInt32, _
 func CGSManagedDisplayGetCurrentSpace(_ cid: CGSConnectionID, _ displayUuid: CFString) -> CGSSpaceID
 
 // adds the provided windows to the provided spaces
-// * macOS 10.10+
+// * macOS 10.10-12.2
```
Comment (lwouis) [inline src/api-wrappers/PrivateApis.swift:125]: RIP 😄

File: src/api-wrappers/CGWindowID.swift [hunk 10]
```diff
@@ -19,11 +19,63 @@ extension CGWindowID {
         return CGSCopySpacesForWindows(cgsMainConnectionId, CGSSpaceMask.all.rawValue, [self] as CFArray) as! [CGSSpaceID]
     }
 
-    func screenshot() -> CGImage? {
-        // CGSHWCaptureWindowList
+    // fullscreen has multiple windows
```
Comment (lwouis) [inline src/api-wrappers/CGWindowID.swift:22]: I don't understand this. I just fullscreen'ed Notes.app, and AltTab shows the same thing I see.

Could you share screenshots to illustrate what you mean by "toolbar window and a main window", and generally what problem this new method fixes?

File: src/logic/Window.swift [hunk 11]
```diff
@@ -83,13 +121,13 @@ class Window {
...
     func refreshThumbnail() {
-        guard let cgImage = cgWindowId.screenshot() else { return }
+        guard let cgImage = cgWindowId.screenshot(self) else { return }
         thumbnail = NSImage(cgImage: cgImage, size: NSSize(width: cgImage.width, height: cgImage.height))
         thumbnailFullSize = thumbnail!.size
     }
 
     func close() {
-        if isWindowlessApp { return }
+        if isWindowlessApp || self.axUiElement == nil { return }
```
Comment (lwouis) [inline src/logic/Window.swift:130]: Does this mean we can't close a window if we haven't visited its Space? That not ideal...

File: src/logic/Window.swift [hunk 12]
```diff
@@ -102,7 +140,7 @@ class Window {
     }
 
     func minDemin() {
-        if isWindowlessApp { return }
+        if isWindowlessApp || self.axUiElement == nil { return }
```
Comment (lwouis) [inline src/logic/Window.swift:143]: See comment above

File: src/logic/Window.swift [hunk 13]
```diff
@@ -119,7 +157,7 @@ class Window {
     }
 
     func toggleFullscreen() {
-        if isWindowlessApp { return }
+        if isWindowlessApp || self.axUiElement == nil { return }
```
Comment (lwouis) [inline src/logic/Window.swift:160]: See comment above

File: src/logic/Window.swift [hunk 14]
```diff
@@ -145,13 +183,43 @@ class Window {
...
+                    BackgroundWork.accessibilityCommandsQueue.asyncWithCap { [weak self] in
+                        guard let self = self else { return }
+                        var psn = ProcessSerialNumber()
+                        GetProcessForPID(self.application.pid, &psn)
+                        _SLPSSetFrontProcessWithOptions(&psn, self.cgWindowId, .userGenerated)
+                        self.makeKeyWindow(psn)
+                        self.axUiElement.focusWindow()
+                    }
+                }
+            } catch {
```
Comment (lwouis) [inline src/logic/Window.swift:220]: I don't think exceptions can bubble up to here. `retryAxCallUntilTimeout` happens on another thread, and absorbs exceptions within

File: src/api-wrappers/CGWindowID.swift (hunk 10 above)
Comment (metacodes) [reply src/api-wrappers/CGWindowID.swift:22]: Before adding the `fullScreenshot` method
image

After adding the `fullScreenshot` method
image

Maybe it depends on machine model. I'm using MBP 16' with Apple Silicon.

File: src/logic/Applications.swift [hunk 15]
```diff
@@ -32,10 +40,11 @@ class Applications {
             let windowsOnOtherSpaces = Spaces.windowsInSpaces(otherSpaces)
             let windowsOnlyOnOtherSpaces = Array(Set(windowsOnOtherSpaces).subtracting(windowsOnCurrentSpace))
             if windowsOnlyOnOtherSpaces.count > 0 {
-                // on initial launch, we use private APIs to bring windows from other spaces into the current space, observe them, then remove them from the current space
-                CGSAddWindowsToSpaces(cgsMainConnectionId, windowsOnlyOnOtherSpaces as NSArray, [Spaces.currentSpaceId])
-                Applications.observeNewWindowsBlocking()
-                CGSRemoveWindowsFromSpaces(cgsMainConnectionId, windowsOnlyOnOtherSpaces as NSArray, [Spaces.currentSpaceId])
+                // Currently we add those window in other space without AXUIElement init
+                // We don't need to get the AXUIElement until we focus these windows.
```
Comment (lwouis) [inline src/logic/Applications.swift:44]: If you're correct that `CGSMoveWindowsToManagedSpace` is a perfect replacement to `CGSAddWindowsToSpaces`, then we could at least fix the 12.3 issue with `CGSAddWindowsToSpaces` not working, but replacing the call to `CGSMoveWindowsToManagedSpace`. That would fix the issue. Then, separately, we can have the conversation about the invisible window trick, and how to move forward there

File: src/api-wrappers/CGWindowID.swift (hunk 10 above)
Comment (lwouis) [reply src/api-wrappers/CGWindowID.swift:22]: Oh, it does it on my machine too (still on Catalina). Very cool then! I wonder how many apps really benefit from that nuance though. It's a cool trick, but in the case of Notes.app, it's not very necessary. Maybe some other apps have an important UI element that would get lost though. I worry a bit about the complexity/performance/potential-failure-points of created the composition vs showing the simple image, though. I think it would probably be worth it to first find an app that truly benefits from this refinement before putting it in. What do you think? Especially given that in order to do this trick, we have to rely on CGS API, which has all the problems I mentioned in other comments

File: src/logic/Applications.swift (hunk 7 above)
Comment (metacodes) [reply src/logic/Applications.swift:133]: I add these codes just to filter those apps which have no windows(e.g. com.apple.universalcontroller). This is for those apps launched before AltTab. At that time, we haven't go to `AXUIElement.isActualWindow`. The code is just there to allow us to avoid observing to applications we are not interested in and to avoid wasting CPU time on `subscribeToNotification` retries.

File: src/api-wrappers/PrivateApis.swift (hunk 8 above)
Comment (metacodes) [reply src/api-wrappers/PrivateApis.swift:131]: This api will move the window from the original space to another space. It means the window will disappear from the original space and show on another space. I did not test the scenarios you mentioned. I just use this api to move the invisible window to the space we want to. I will do some tests. By the way, I can not do the tests for multiple monitors, because I only have one monitor.😂

File: src/api-wrappers/CGWindowID.swift (hunk 10 above)
Comment (metacodes) [reply src/api-wrappers/CGWindowID.swift:22]: Indeed, I also took into account the performance issues. In fact, this small issue has little impact, I just made it a little bit optimized as free time. Reminders.app, Hopper Disassembler.app, Xcode.app, Microsoft Word.app, Calendar.app also have this problem as I known.

Comment (metacodes) [reply src/api-wrappers/CGWindowID.swift:22]: Initially I tried to use CGWindowListCreateImageFromArray, but it didn't work at all. Maybe I'm not using it the right way.

File: src/logic/Window.swift (hunk 11 above)
Comment (metacodes) [reply src/logic/Window.swift:130]: It's a pity that HyperSwitch also can't do that either.

Comment (lwouis) [reply src/logic/Window.swift:130]: I just tested and it can.

I did this to test:

* Opened a new window
* Send it to another Space
* Launch HyperSwitch
* Press alt-tab
* Notice that HS shows the window
* Hover the mouse over the window in HS
* A close button appears, press it
* The window on the other Space is closer, even though I never navigated to that Space

How can HS close the window without the AX reference?

**Comment by lwouis** [comment]
FYI, I just pushed a release with these 2 commits: https://github.com/lwouis/alt-tab-macos/commit/0569ed0f8603ca0611c9d6bc0f00046f48c4415e https://github.com/lwouis/alt-tab-macos/commit/cdb461ab6820b21146ee16fa85d197b846e1b276

I thought they may mitigate the high CPU issues people are facing at the moment. I think this buys more time for this pull-request to become more robust and maybe deliver breakthroughs

File: src/logic/Window.swift (hunk 11 above)
Comment (metacodes) [reply src/logic/Window.swift:130]: On macOS 12.3.1, I did the steps same to you, but HyperSwitch just took me to that space and didn't close the window. Maybe the private APIs they used also broken.

File: src/logic/Applications.swift (hunk 15 above)
Comment (lwouis) [reply src/logic/Applications.swift:44]: Please see my follow-up here: https://github.com/lwouis/alt-tab-macos/issues/1324#issuecomment-1120156153

**Comment by lwouis** [comment]
@metacodes I'll close this PR now since I merged part of it. If you're still interested in improving perf, could you please open a new PR with your new improvements?

**Comment by metacodes** [comment]
> @metacodes I'll close this PR now since I merged part of it. If you're still interested in improving perf, could you please open a new PR with your new improvements?

OK👌
